### PR TITLE
feat: Improve credential support for URL signing.

### LINKED ISSUE #8413: Improve how Cloud Storage presigned requests are generated
Currently, pre-signing requests for cloud storage is - for lack of better words - [a bit painful](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Storage.V1/latest/index#signing-urls-without-a-service-account-credential-file).

I would love to see it as simple as:

```csharp
var credential = (await GoogleCredential.GetDefaultServiceAccount()) ?? throw new("No service account assigned.");
var urlSigner = await UrlSigner.FromServiceAccountCredential(credential);
var uri = await urlSigner.SignAsync(someBucket, key, duration, method);
```

The key focus here is that I think the APIs could afford to be a little more discoverable and self-evident in terms of how a user can obtain whatever credential type that satisfies their situation. The most crucial aspect being that right now, service accounts feel kind of second-class, which inhibits the convenience of being able to assign them to things like Google Cloud Run revisions . :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by amanda-tarafa**
Closes #8413

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
Draft for now, because snippet changes and possibly tests are missing.
Also the companion Google.Apis.Auth PR will be coming after, but that only has to make ComputeCredential an IBlobSigner.
This is mostly to agree on surface, etc.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:164]: Note that if you want to leave the TODO and deprecate it later, the comment will need to move.

**Review by jskeet — COMMENTED** [review]
Yup, this looks good.

**Comment by amanda-tarafa** [comment]
Requirement PR: https://github.com/googleapis/google-api-dotnet-client/pull/2308

**Comment by amanda-tarafa** [comment]
@jskeet this now includes all method deprecations and alternatives. PTAL just before I make any changes in tests due to deprecation, as those are a lot.
Also, some factory methods are now async given that the Compute service account ID fetch operation is async.

(Of course this PR won't build until we have merged and release https://github.com/googleapis/google-api-dotnet-client/pull/2308)

**Comment by amanda-tarafa** [comment]
And ignore the first commit marked `temp`.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:150]: Possibly just FromCredentialFileAsync? That would be consistent with GoogleCredential.FromFileAsync.

I think it's fine for *this* method to be async, btw, given that the asynchronous part is loading the credential.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:127]: Does this definitely need to be asynchronous? Can't we wait until the blob signing operation (which is already asynchronous) to get the target principal?

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:159]: FromCredentialStream maybe? (Again, to be consistent with GoogleCredential.FromStream)

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.CredentialBlobSigner.cs
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.CredentialBlobSigner.cs:31]: I think if this took a `Func<Task<string>>` as an asynchronous ID provider, we could avoid some factory methods being async.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs
Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs]: Yep, done

**Review by amanda-tarafa — COMMENTED** [review]
Comments addressed and new approach implemented (save for tests).

@jskeet PTAL, second and third commits only.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:207]: For some reason I can't leave a suggestion, but: "When called, must not return a null reference or a task which completes with a null result."

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs
Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:209]: Yep, on both, changing the name to provider and making the method private.

**Review by amanda-tarafa — COMMENTED** [review]
All comments addressed, changes will be pushed shortly.

File: apis/Google.Cloud.Storage.V1/docs/index.md
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/docs/index.md:60]: signig => signing

File: apis/Google.Cloud.Storage.V1/docs/index.md
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/docs/index.md:66]: credentials => credential

File: apis/Google.Cloud.Storage.V1/docs/index.md
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/docs/index.md:71]: I think "take care" is probably a little too strong. How about "it's worth being aware of the underlying synchronous/asynchronous nature" or something similar. (And I'd expand "sync" to "synchronous" everywhere, although "async" is probably okay.)

**Review by jskeet — APPROVED** [review]
A few comments on the documentation, but that's all.

I'm slightly nervous about the use of Task.Run in the signer factory, but we can revisit that later if we want to (as everything is private).

**Comment by amanda-tarafa** [comment]
>I'm slightly nervous about the use of Task.Run in the signer factory, but we can revisit that later if we want to (as everything is private).

@jskeet see the last commit and let me know if that looks better. I haven't addressed the rest of your comments yet, I'll do that before merging if you are OK with this one.

**Review by jskeet — APPROVED** [review]
Happy to merge this (once doc niggles are done), but as discussed offline we're going to wait until next week to release, in case we think of better options.

**Pull request merged**
